### DISCUSSION TITLE
rdctl set for Wsl Integration

### DISCUSSION THREAD

**Discussion by SamambaMan** [rancher-sandbox/rancher-desktop#6442] [Q&A] [answered]
Hello there!

I'm trying to automate a Rancher installation and one of the steps would be activate the WSL integration for an Ubuntu WSL installation, but failing super hard.

I'm trying to use several combinations of  "rdctl set --application.WSL.integration"..., but no success.

If I list `rdctl list-settings` there is an entry for WSL: 
```json  
"WSL": {
    "integrations": {
      "Ubuntu-20.04": true,
      "Ubuntu": true
    }
  },
```

But if I do something like:
```powershell
 rdctl api /v1/settings --method PUT --body '{"WSL": {"integrations": {"Ubuntu-20.04": true}}}'
```
I get in return
```
error processing JSON request block
{"message":"400 Bad Request"}
```

But trying to reproduce with the example in the documentation:
https://docs.rancherdesktop.io/references/rdctl-command-reference/#rdctl-api-v1settings

```powershell
rdctl api /v1/settings --method PUT --body '{"kubernetes": {"enabled": false}}'
```

Gives me the same 400 error.

Do I have to enable the API somehow somewhere?

Thanks in advance!

**Comment by mook-as** [comment]
Hi! If you're using PowerShell 5.1, there's some weird quoting involved that I haven't figured out. (You can check this if you look at `server.log`; I don't think debug mode is required.)

I get something like this:
```
2024-02-07T18:17:11.513Z: updateSettings: error processing JSON request block
{WSL: {integrations: {Ubuntu-20.04: true}}}
 SyntaxError: Expected property name or '}' in JSON at position 1
```
Notice that all the quotes are missing, so it's no longer valid JSON.
Try something like:

```ps
rdctl api /v1/settings --method PUT --body '{\"WSL\": {\"integrations\": {\"Ubuntu-20.04\": true}}}'
```

Note that this does not appear to be the case with PowerShell 7.x (the one installed from the MS Store).

**Reply by SamambaMan** [reply]
Hey, thank you for the super quick response!

Indeed, what you say seems relating, but I don't get the same behavior.
When running that way I get the following logs: 
```
2024-02-07T18:36:26.519Z: updateSettings: error processing JSON request block
{\"WSL\": {\"integrations\": {\"Ubuntu-20.04\": true}}}
 SyntaxError: Expected property name or '}' in JSON at position 1
    at JSON.parse (<anonymous>)
    at Ji.readRequestSettings (C:\Program Files\Rancher Desktop\resources\app.asar\dist\app\background.js:41:80633)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Ji.updateSettings (C:\Program Files\Rancher Desktop\resources\app.asar\dist\app\background.js:41:80905)
```

I will debug this for a few minutes, but thanks for so kindly tell me "read the F logs!" hahaha thanks

I'm indeed on powershell 5.1:
```
Name                           Value
----                           -----
PSVersion                      5.1.22621.2506
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.22621.2506
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
```

Will keep trying

**Reply by mook-as** [reply, marked as answer]
Oh! Interesting; it looks like for you the backslashes are sent literally (somehow), so you need to _not_ escape it. It's probably best to check the logs from your original input, then.

I don't understand how behaviour is different between the two PowerShell 5.1s, but it looks like you're probably on Windows 11 (I have 5.1.19041.3930 on Windows 10), so maybe that? Or there's a config difference somewhere.  I copied your input directly, so that's really confusing…

(Note that for recent Rancher Desktop you'd also need to supply a `version` field, but the error message should tell you that if that's the case.)

**Reply by SamambaMan** [reply]
Oh, nevermind!

It was just me messing with the terminals.
Restarted and went just fine.

Got the "version" error you mentioned, fixed right away, works fine now! 

```
PS D:\Users\FFerreira> rdctl api /v1/settings --method PUT --body '{\"version\": 1, \"WSL\": {\"integrations\": {\"Ubuntu-20.04\": true}}}'
settings updated; no restart required
PS D:\Users\FFerreira>
```

That was awesome, thanks!
